hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java

```diff
@@ -228,63 +674,436 @@ private void unblockAll(Object waiter, Executor executor) {
             if (waiter instanceof Thread) {
                 unpark((Thread) waiter);
                 return;
-            } else if (waiter instanceof ExecutionCallback) {
-                unblock((ExecutionCallback) waiter, executor);
-                return;
             } else if (waiter.getClass() == WaitNode.class) {
                 WaitNode waitNode = (WaitNode) waiter;
                 unblockAll(waitNode.waiter, waitNode.executor);
                 waiter = waitNode.next;
             } else {
+                unblockOtherNode(waiter, executor);
                 return;
             }
         }
     }
 
-    private void unblock(final ExecutionCallback<V> callback, Executor executor) {
+    private CompletableFuture<Void> unblockAccept(@Nonnull final Consumer<? super V> consumer,
+                                                    @Nonnull Executor executor,
+                                                    @Nonnull CompletableFuture<Void> future) {
+        final Object value = resolve(state);
+        if (cascadeException(value, future)) {
+            return future;
+        }
         try {
-            executor.execute(new Runnable() {
-                @Override
-                public void run() {
-                    try {
-                        Object value = resolve(state);
-                        if (value instanceof Throwable) {
-                            Throwable error = unwrap((Throwable) value);
-                            callback.onFailure(error);
-                        } else {
-                            callback.onResponse((V) value);
-                        }
-                    } catch (Throwable cause) {
-                        logger.severe("Failed asynchronous execution of execution callback: " + callback
-                                + " for call " + invocationToString() + ": " + cause, cause);
-                    }
+            executor.execute(() -> {
+                try {
+                    consumer.accept((V) value);
+                    future.complete(null);
+                } catch (Throwable t) {
+                    future.completeExceptionally(t);
                 }
-
             });
         } catch (RejectedExecutionException e) {
-            callback.onFailure(wrapToInstanceNotActiveException(e));
+            future.completeExceptionally(wrapToInstanceNotActiveException(e));
+        }
+        return future;
+    }
+
+    /**
+     *
+     * @param waiter    the current wait node, see javadoc of {@link #state state field}
+     * @param executor  the {@link Executor} on which to execute the action associated with {@code waiter}
+     */
+    @SuppressWarnings("checkstyle:CyclomaticComplexity")
+    protected void unblockOtherNode(Object waiter, Executor executor) {
```

Comment:
Looks like a perfect place to introduce a virtual method for `execute(executor, value)` pattern, but probably there is a reason not to.